### PR TITLE
[backport: 20.09] darwin (Big Sur): tbd based stdenv

### PR DESCRIPTION
**Pull request by davidtwco**
###### Motivation for this change
This PR backports #98541 to 20.09. This is helpful for projects using NixOps that have a pinned nixpkgs commit to a stable NixOS release, in that circumstance, developers using Big Sur can't enter the `nix-shell` without these changes.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
While I was working on pyobjc I needed the two commits below to to build some things. I am not sure we should backport such major change to a stable version. Big Sur has probably more bugs and telling them to use the unstable channel until 21.XX is out is probably less hassle.

https://github.com/NixOS/nixpkgs/pull/102306/commits/79f3fa30a07efd72a4c3f2cac2fce4f9e2350c91
https://github.com/NixOS/nixpkgs/pull/102306/commits/0cc95f6b1937383118507c6ae549d210f96cea75

**Comment by davidtwco** [comment]
> I am not sure we should backport such major change to a stable version. Big Sur has probably more bugs and telling them to use the unstable channel until 21.XX is out is probably less hassle.

I think there's enough value in having a stable branch which can be pinned by projects and have those projects still be usable on Big Sur to justify this backport.

**Comment by matthewbauer** [comment]
We usually don't backport big stdenv changes like this. However, since Big Sur is pretty much broken without it, I think it's probably a good idea in this case.

**Comment by LnL7** [comment]
I would definitively like to backport the fixes, we can test for regressions with a wip jobset to make sure nothing important was missed.  However I think it would be best to wait until bootstrapping and some of the missing stubs are sorted out.

**Comment by chessai** [comment]
Agreed, Big Sur breakage is too big a deal to not backport IMO

**Comment by jonringer** [comment]
If someone is able to verify that there's no regressions, I'm fine with merging this. But it should target `staging-20.09`

**Comment by davidtwco** [comment]
I've rebased this onto `staging-20.09` as requested, and included #107013.

**Comment by phile314** [comment]
Is there anything that can be done to help getting this merged? If there is anything to test, how do I best test it?

**Comment by jonringer** [comment]
> Is there anything that can be done to help getting this merged? If there is anything to test, how do I best test it?

building the major packages would be a good start, but as I don't have access to darwin hardware (and doing anything else is technically illegal), it's up to @NixOS/darwin-maintainers  to merge this.

**Comment by jonringer** [comment]
@SuperSandro2000 if you could build some of the more major packages, I'm assuming this is good as unstable seems to be fine. However, we could be missing a PR which is correcting unstable behavior.

File: pkgs/development/compilers/llvm/10/compiler-rt.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/llvm/10/compiler-rt.nix:44]:
```suggestion
  ] ++ lib.optionals (stdenv.hostPlatform.isDarwin) [
```

stdenv.lib is deprecated.

File: pkgs/development/compilers/llvm/7/libc++abi.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/llvm/7/libc++abi.nix:26]:
```suggestion
     lib.optional standalone "-DLLVM_ENABLE_LIBCXX=ON" ++
     lib.optional (!enableShared) "-DLIBCXXABI_ENABLE_SHARED=OFF";
```

File: pkgs/os-specific/darwin/apple-sdk/print-reexports/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/darwin/apple-sdk/print-reexports/default.nix:5]:
```suggestion
  src = lib.sourceFilesBySuffices ./. [".c"];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 if you could build some of the more major packages, I'm assuming this is good as unstable seems to be fine. However, we could be missing a PR which is correcting unstable behavior.

I am against backporting any of these major stdenv changes to 20.09. If you are using Big Sur you either need to wait for 21.05 or use unstable until then. Any of those changes could break a lot and the Darwin maintainers are rare currently so I don't want to deal with any of the potential bugs.

I can try to build a few packages but rebuilding stdenv never worked for me so I expect it to fail.

**Comment by domenkozar** [comment]
Note that a lot of macOS users do update their OS and this completely breaks their system currently on 20.09, so having fewer packages to work is better than zero.

**Comment by jonringer** [comment]
if any bigsur user wants to help this along, you can run a nixpkgs review of all the channel blockers and a package from most of the toolchains

```
nixpkgs-review pr 105799 \
  --package ripgrep \
  --package python3Packages.requests \
  --package direnv \
  --package vim \
  --package vim \
  --package vim \
  --package vim \
  --package vim \
  --package vim \

**Comment by SuperSandro2000** [comment]
> if any bigsur user wants to help this along, you can run a nixpkgs review of all the channel blockers and a package from most of the toolchains

The problem with this is that nixpkgs-review activates the sandbox which prevents rebuilds of stdenv on darwin.

**Comment by jonringer** [comment]
> The problem with this is that nixpkgs-review activates the sandbox which prevents rebuilds of stdenv on darwin.

Oh.... that's awkward. And I was not aware of this.

Guess I'm just having it easy in purity land.

**Comment by MarcoPolo** [comment]
Is there a way to disable the sandbox?

I have a Big Sur machine handy. Anything I can do to help?

**Comment by SuperSandro2000** [comment]
> Is there a way to disable the sandbox?

For nixpkgs-review. Not currently but I am talking to @Mic92 to solve this.

> I have a Big Sur machine handy. Anything I can do to help?

You can try building this PR and see if it even works.

**Comment by SuperSandro2000** [comment]
Tried building python3:

```
checking build system type... x86_64-apple-darwin19.6.0
checking host system type... x86_64-apple-darwin19.6.0
checking for a BSD-compatible install... /nix/store/ai232lgqxzkhwl0zd3pvprj3hdgab6v6-coreutils-8.31/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /nix/store/ai232lgqxzkhwl0zd3pvprj3hdgab6v6-coreutils-8.31/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports the include directive... yes (GNU style)
checking for gcc... clang
checking whether the C compiler works... no
configure: error: in `/private/var/folders/r4/k85s52jx3555ckj0wty615s00000gp/T/nix-build-expat-2.2.8.drv-0/expat-2.2.8':
configure: error: C compiler cannot create executables
See `config.log' for more details
builder for '/nix/store/w4yqg2rch50rqpwqh1pz94863hbzi0iv-expat-2.2.8.drv' failed with exit code 77
cannot build derivation '/nix/store/bd1523xydp5mndb0d5qk8qw6vnsf8mbc-python3-3.8.5.drv': 1 dependencies couldn't be built
error: build of '/nix/store/bd1523xydp5mndb0d5qk8qw6vnsf8mbc-python3-3.8.5.drv' failed
```

**Comment by MarcoPolo** [comment]
Yep, failed at the same spot as Sandro on Big Sur.

**Pull request closed without merging**
